db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -4106,9 +4114,15 @@ Status DBImpl::WaitForCompact(
     }
     if ((bg_bottom_compaction_scheduled_ || bg_compaction_scheduled_ ||
          bg_flush_scheduled_ || unscheduled_compactions_ ||
-         unscheduled_flushes_) &&
+         unscheduled_flushes_ || error_handler_.IsRecoveryInProgress()) &&
         (error_handler_.GetBGError().ok())) {
       bg_cv_.Wait();
+    } else if (wait_for_compact_options.close_db) {
+      shutting_down_.store(true, std::memory_order_release);
```

Comment:
The impact of this may be beyond disallowing more background work though. For example, there is logic in `CancelAllBackgroundWork` that checks for flush based on this flag: https://github.com/facebook/rocksdb/blob/main/db/db_impl/db_impl.cc#L510-L527

Not sure how that interacts with the effect of the `wait_for_compact_options.flush` option above.

Reply:
Yup, this is why I added the flush logic from L4098 to L4105. The lines you are referencing should be equivalent to `DBImpl::FlushAllColumnFamilies(FlushOptions(), FlushReason::kShutDown);`

This might change the behavior slightly when `close_db` option is on. Flush for unpersisted data will be handled as the following before we wait for compaction. The unpersisted data flush in `CancelAllBackgroundWork()` will be skipped.

1. If flush option is on, flush for unpersisted data will be handled by manual flush triggered in line 4093.
2. If flush option is off, flush for unpersisted data will be handled with FlushReason::kShutDown in line 4102 (only when the `mutable_db_options_.avoid_flush_during_shutdown` is false)